ipcache: Add ExternalIP of the local node to ipcache

AnishShah opened the pull request:
Please ensure your pull request adheres to the following guidelines:

- [x] For first time contributors, read [Submitting a pull request](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#submitting-a-pull-request)
- [x] All code is covered by unit and/or runtime tests where feasible.
- [x] All commits contain a well written commit description including a title,
      description and a `Fixes: #XXX` line if the commit addresses a particular
      GitHub issue.
- [x] All commits are signed off. See the section [Developer’s Certificate of Origin](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#dev-coo)
- [x] Provide a title or release-note blurb suitable for the release notes.
- [x] Thanks for contributing!

ExternalIP of the local k8s node is not added to the IPCache. So when a client inside the cluster tries to connect to a NodePort Service via ExternalIP of the local k8s node, no service resolution happens.

Fixes: #14765 

```release-note
Fix a bug that affects connectivity to NodePort service via ExternalIP of the local k8s node.
```

AnishShah committed:
Rename ExternalIPv4 to IPv4 and InternalIPv4 to InternalIPv4Router

[SG]etInternalIPv4Router - This is the CiliumInternalIP of the node. It
is only relevant within the Cilium-managed network (this means within
the node for direct routing mode and on the overlay for tunnel mode).
Inorder to not get confused with InternalIP of a k8s node, we renamed
`[SG]etInternalIPv4` to `[SG]etInternalIPv4Router`.

[SG]etIPv4 - This returns one of the IPv4 node address based on the
following priority:
- NodeInternalIP
- NodeExternalIP
- other IP address type.
These are reachable on the network. Since it can be External or
Internal IP address, we renamed `[SG]etExternalIPv4` to `[SG]etIPv4`

Signed-off-by: Anish Shah <anishshah@google.com>

AnishShah committed:
Store ExternalIP of a k8s node so that it can be added to IPCache

Previously, accessing a NodePort using ExternalIP of the local K8s node
would result in connection timeout. This is because ExternalIP of the
local K8s node is not added to ipcache and hence no service resolution
happens on egress.

This commit resolves this issue by adding the ExternalIP of local k8s
node to the ipcache. We introduce two functions `SetExternalIPv[46]` and
`GetExternalIPv[46]` to store/fetch the External IPv4/IPv6 address of a k8s node.
If a node doesn't have any External IP address, then we store/return
`nil`.

Signed-off-by: Anish Shah <anishshah@google.com>

brb commented on pkg/node/address.go:
```diff
@@ -236,6 +237,16 @@ func GetIPv4Router() net.IP {
 }
 
+// SetExternalIP sets the external IP node address. It must be a public IP.
+func SetExternalIP(ip net.IP) {
```
I think we should rename this to `SetK8sExternalIP` (same goes to `GetExternalIP`) to better reflect that this is k8s Node ExternalIP rather than a public facing iface IP addr?

Also, what happens in the DualStack setup? Is it possible to have different IP addrs for IPv4 and IPv6? /cc @fristonio

AnishShah replied:
Renamed it to `SetExternalIPv4` and `SetExternalIPv6`. I think in DualStack setup, it is possible to have different IP address for IPV4 and IPV6.

brb commented on pkg/node/address.go:
```diff
@@ -46,6 +46,7 @@ var (
 	ipv6NodePortAddrs map[string]net.IP // iface name => ip addr
 	ipv4AllocRange    *cidr.CIDR
 	ipv6AllocRange    *cidr.CIDR
+	externalAddress   net.IP
```
Could you add a comment saying that this is k8s Node ExternalIP?

AnishShah replied:
Done.

brb reviewed (changes requested):
Thanks for the PR. Some nits.

Also, I'm not very familiar with the IPCache subsystem. Therefore we need to get a review from @cilium/ipcache

joestringer commented on pkg/node/address.go:
```diff
@@ -216,30 +216,30 @@ func GetIPv6AllocRange() *cidr.CIDR {
-func GetInternalIPv4() net.IP {
-	return ipv4InternalAddress
+// GetIPv4Router returns the cilium internal IPv4 node address
+func GetIPv4Router() net.IP {
```
Maybe, `[SG]etInternalIPv4Router()`?

I believe that the intent behind the previous "InternalIPv4()" naming was not to conflate with k8s internal IPs, but instead to hint that this IP address is only relevant internally within the Cilium-managed network (or in the case of direct routing, only within the node). AFAIK Cilium (or the Linux stack) won't respond to ARP for this IP on the native devices, only on the `cilium_host` device. It'd be nice if we could retain this context in the naming and/or godoc here.

AnishShah replied:
Done. Added this information in the commit as well as godoc.

joestringer commented on pkg/k8s/init.go:
```diff
@@ -245,6 +245,9 @@ func WaitForNodeInformation() error {
 
 		node.SetLabels(n.Labels)
 
+		node.SetK8sExternalIPv4(n.GetExternalIP(false))
```
The commit message here says:

> NodePort is not accessible via ExternalIP of the local node. This is
because the ExternalIP of the local node is not added to the IPCache and
hence service resolution does not happen.

The above does not describe whether this was the behaviour before or after this change, and therefore lacks the declaration of intent of what is trying to be achieved. This means it's up to the reader to try to figure out whether you are trying to define what the problem was (and hence what this commit fixes), or what the intended behaviour is (in which case it should match exactly what the code says).

Please consider describing the intent with broader context inside the commit message, roughly:

* *Previously,* accessing a NodePort on the ExternalIP of the node would result in `<behaviour>`. This is because `<reason>`.
* This commit changes/resolves this by adding the local ExternalIP to the ipcache.

AnishShah replied:
Done. I also added information about `[SG]etK8sExternalIPv4` to the commit.

joestringer commented on pkg/node/address.go:
```diff
@@ -236,6 +240,16 @@ func GetIPv4Router() net.IP {
+// GetK8sExternalIPv4 returns the external IPv4 node address.
+func GetK8sExternalIPv4() net.IP {
+	return ipv4ExternalAddress
+}
```
What's the semantic difference between the `GetIPv4()` and `GetK8sExternalIPv4()`? How does someone in future decide which one to use? Are they really different IP addresses?

AnishShah replied:
Added details to the godoc as well as the commit. Basically,

`GetK8sExternalIPv4()` is strictly an IPv4 address that is routable on the public internet.

`GetIPv4()` can be one of the address with the following priority:
- `NodeInternalIP`
- `NodeExternalIP`
- other IP address types.

joestringer commented on pkg/k8s/init.go:
```diff
@@ -245,6 +245,9 @@ func WaitForNodeInformation() error {
 		node.SetLabels(n.Labels)
 
+		node.SetK8sExternalIPv4(n.GetExternalIP(false))
+		node.SetK8sExternalIPv6(n.GetExternalIP(true))
```
Did you intend to do an `if option.Config.EnableIPv4` /  `if option.Config.EnableIPv6`  check here anywhere? At face value I would assume that if the IP protocol is disabled, we should not be attempting to configure that particular IP.

AnishShah replied:
I didn't do `if option.Config.EnableIPv4 / if option.Config.EnableIPv6` check here because of the comment on line 232.

https://github.com/cilium/cilium/blob/0c760d8c5b4437e1017eb0ef9c0a74f84e68a6d4/pkg/k8s/init.go#L232-L237

AnishShah replied:
Should I still add it?

joestringer replied:
In that case, maybe not. I presume that these external IPs would be suitable for use from clustermesh given that the requirement is they're publicly routable (though obviously this code is not making such a change as yet).

joestringer reviewed (changes requested):
Hi @anishshah, thanks for the submission.

The core fix to add the IP to the IPcache makes sense to me. Given that other nodes already have corresponding IPcache entries for the ExternalIPs, I don't think that there are other regular IPcache concerns (such as how it may affect policy, for instance).

I'm a little concerned that we're making the IP address configuration even more confusing in the codebase by renaming & adding yet another (maybe different? maybe not?) IP address to deal with in the node package. It would be helpful for review if there was a description with an overview of the addresses involved in configuration, ideally with examples from a demo cluster. Some of my questions/comments in review below are trying to drive at this question.

This PR also fixes a bug, but there is no accompanying integration test. Logically the argument seems correct, but without a test it's hard to guarantee that this change fixes the issue, and also in future without a test Cilium could easily regress. Maybe @brb has an idea how to bootstrap a simple test to verify this fix (eg, a similar test that this could be based upon)?

joestringer commented on pkg/node/address.go:
```diff
@@ -216,30 +220,55 @@ func GetIPv6AllocRange() *cidr.CIDR {
+
+// SetInternalIPv4Router sets the cilium internal IPv4 node address, it is allocated from the node prefix.
+// This must not be conflated with k8s internal IP as this IP address is only relevant within the node.
+// Cilium doesn't respond to ARP for this IP on the native devices, but only on the `cilium_host` device.
```
While I believe that this statement is correct today, I don't think it's a good idea to document it here since that behaviour would be subject to entirely different code so this comment could easily be wrong or get out of date as changes are made to the codebase.

```suggestion
```

AnishShah replied:
Fixed.

joestringer commented on pkg/node/address.go:
```diff
@@ -216,30 +220,55 @@ func GetIPv6AllocRange() *cidr.CIDR {
+}
+
+// SetInternalIPv4Router sets the cilium internal IPv4 node address, it is allocated from the node prefix.
+// This must not be conflated with k8s internal IP as this IP address is only relevant within the node.
```
I don't think that `only relevant within the node` is quite correct; in some situations, this IP may be used as the IP for host -> remote pod traffic. I would use the terms `only relevant within the Cilium-managed network`.

We could potentially further qualify it by saying that this means "within the node" for direct routing mode or "on the overlay" for tunnel mode, though that might be bordering on too much detail...? :thinking:

AnishShah replied:
Fixed.

joestringer reviewed (changes requested):
Couple of other minor nits on the new comments. Otherwise it seems fine to me.

AnishShah committed:
test: Add test to check NodePort service connectivity from inside the cluster

This test checks connectivity to NodePort service from inside a cluster
by sending request to ExternalIP:NodePort.

Signed-off-by: Anish Shah <anishshah@google.com>

brb commented on test/k8sT/Services.go:
```diff
@@ -1124,6 +1130,20 @@ Secondary Interface %s :: IPv4: (%s, %s), IPv6: (%s, %s)`, helpers.DualStackSupp
 			}
 
 			if helpers.RunsOnGKE() {
+				if bpfNodePort && helpers.RunsWithoutKubeProxy() {
```
Let's remove the `bpfNodePort && helpers.RunsWithoutKubeProxy()` condition. Otherwise it won't be tested on the current GKE CI job (see https://github.com/cilium/cilium/pull/14793#issuecomment-772379577).

AnishShah replied:
Fixed.

brb commented on test/k8sT/Services.go:
```diff
@@ -93,8 +95,12 @@ var _ = Describe("K8sServicesTest", func() {
 
-		k8s1NodeName, k8s1IP = kubectl.GetNodeInfo(helpers.K8s1)
-		k8s2NodeName, k8s2IP = kubectl.GetNodeInfo(helpers.K8s2)
+		k8s1NodeName, k8s1InternalIP = kubectl.GetNodeInfo(helpers.K8s1)
```
Nit: can we make `GetNodeInfo` to return both IP addrs?

AnishShah replied:
Fixed.

nebril commented on test/helpers/kubectl.go:
```diff
@@ -3654,12 +3654,13 @@ func (kub *Kubectl) GetCiliumPodOnNode(node string) (string, error) {
-	return nodeName, nodeIP
+	nodeInternalIP, err = kub.GetNodeIPByLabel(label, false)
+	gomega.ExpectWithOffset(1, err).Should(gomega.BeNil(), "Can not retrieve Node Internal IP for "+label)
+	nodeExternalIP, _ = kub.GetNodeIPByLabel(label, true)
```
If this ever fails it will be pretty hard to figure it out. Please handle the error with `gomega.ExpectWithOffset(1, err).Should(gomega.BeNil(), "Can not retrieve Node External IP for "+label)`, same as is done for internal IP.

AnishShah replied:
I'm not sure if we should fail in case we can't retrieve ExternalIP. Does all the integration tests include kubernetes node with an external IP?

AnishShah replied:
We return an error if a node does not have any external IP assigned. If all k8s nodes in our integration test has an ExternalIP, then it is safe to fail here. Otherwise this should be fine.

https://github.com/cilium/cilium/blob/e567f8bfa7e8cc4d71566f224f60d7222d3d874b/test/helpers/kubectl.go#L1082-L1100

nebril replied:
These tests can be run on any k8s cluster that is set up to run them ( https://docs.cilium.io/en/v1.9/contributing/testing/e2e/#running-end-to-end-tests-in-other-environments-via-kubeconfig ), lack of handling this error potentially costs a developer debugging the test on a cluster without external IPs non-trivial amount of time. If we just fail the test here saying "there are no external IPs and there should be", we give the developer a clear message on what is wrong with their setup.

AnishShah replied:
The issue is `GetNodeInfo` function is used in many tests including ones that don't need k8s node with externalIPs. So my question was whether all nodes in CIlium CI jobs have external IPs? If not, then the tests will fail even if they didn't need k8s node with externalIPs.

nebril replied:
Ah, I see what you mean. In this case, we should log a warning that the external IP was not retrieved, silently ignoring errors will lead to a lot of frustrations.

AnishShah replied:
Thanks. I changed things a bit. I try to retrieve ExternalIP of the Node only when we run on GKE as that's when we need it. I fail the test if we can't retrieve it.

nebril reviewed (changes requested):
Thanks for the PR! There is one issue that I think requires fixing in `GetNodeInfo` error handling.

christarazi commented on test/k8sT/Services.go:
```diff
@@ -1124,6 +1124,17 @@ Secondary Interface %s :: IPv4: (%s, %s), IPv6: (%s, %s)`, helpers.DualStackSupp
 
 			if helpers.RunsOnGKE() {
+				k8s1ExternalIP, err := kubectl.GetNodeIPByLabel(helpers.K8s1, true)
+				Expect(err).Should(BeNil(), "Cannot retrieve Node External IP for"+helpers.K8s1)
```
`"Cannot retrieve Node External IP for"+helpers.K8s1`

There's no space between `for` and `helpers.K8s1`, could you add spacing in between there?

Also, `Should*` accepts a format string as the second argument, and additional args. For ex:

```
Expect(err).Should(BeNil(), "Cannot retrieve Node External IP for %s", helpers.K8s1)
```

This should be preferred.

AnishShah replied:
Fixed.

christarazi commented on test/k8sT/Services.go:
```diff
@@ -1124,6 +1124,17 @@ Secondary Interface %s :: IPv4: (%s, %s), IPv6: (%s, %s)`, helpers.DualStackSupp
+				k8s1ExternalIP, err := kubectl.GetNodeIPByLabel(helpers.K8s1, true)
+				Expect(err).Should(BeNil(), "Cannot retrieve Node External IP for"+helpers.K8s1)
+				k8s2ExternalIP, err := kubectl.GetNodeIPByLabel(helpers.K8s2, true)
+				Expect(err).Should(BeNil(), "Cannot retrieve Node External IP for"+helpers.K8s2)
```
Same here

AnishShah replied:
Fixed.

christarazi reviewed (changes requested):
Thanks for the PR, overall looks good! A few minor nits regarding the e2e test commit:

- See below comments
- Could you rename the commit to `test: Add test to check NodePort service connectivity from inside cluster`? This follows the pattern that exists in our [developer guide](https://docs.cilium.io/en/latest/contributing/development/contributing_guide/#submitting-a-pull-request).
- I noticed that you reference the issue in each commit, even though technically only the first two are needed in order to fix the issue. Therefore, I think it makes more sense to mention the `Fixes` line in the PR body itself and remove them from the individual commits.

christarazi reviewed (approved):
:tada:

joestringer reviewed (approved):
LGTM.
